crates/bevy_ui/src/render/ui_material_pipeline.rs

```diff
@@ -0,0 +1,779 @@
+};
+use bevy_transform::prelude::GlobalTransform;
+use bevy_utils::{FloatOrd, HashMap, HashSet};
+use bevy_window::{PrimaryWindow, Window};
+use bytemuck::{Pod, Zeroable};
+
+use crate::{
+    resolve_border_thickness, CalculatedClip, Node, RenderUiSystem, Style, TransparentUi,
+    UiMaterial, UiMaterialKey, UiScale, UiStack, QUAD_INDICES, QUAD_VERTEX_POSITIONS,
+};
+
+pub const MATERIAL_UI_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10074188772096983955);
+
+const UI_VERTEX_OUTPUT_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10123618247720234751);
+
+/// Adds the necessary ECS resources and render logic to enable rendering entities using the given
+/// [`UiMaterial`] asset type (which includes [`UiMaterial`] types).
+pub struct UiMaterialPlugin<M: UiMaterial>(PhantomData<M>);
+
+impl<M: UiMaterial> Default for UiMaterialPlugin<M> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<M: UiMaterial> Plugin for UiMaterialPlugin<M>
+where
+    M::Data: PartialEq + Eq + Hash + Clone,
+{
+    fn build(&self, app: &mut bevy_app::App) {
+        load_internal_asset!(
+            app,
+            UI_VERTEX_OUTPUT_SHADER_HANDLE,
+            "ui_vertex_output.wgsl",
+            Shader::from_wgsl
+        );
+        load_internal_asset!(
+            app,
+            MATERIAL_UI_SHADER_HANDLE,
+            "ui_material.wgsl",
+            Shader::from_wgsl
+        );
+        app.init_asset::<M>()
+            .add_plugins(ExtractComponentPlugin::<Handle<M>>::extract_visible());
+
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app
+                .add_render_command::<TransparentUi, DrawUiMaterial<M>>()
+                .init_resource::<ExtractedUiMaterials<M>>()
+                .init_resource::<ExtractedUiMaterialNodes<M>>()
+                .init_resource::<RenderUiMaterials<M>>()
+                .init_resource::<UiMaterialMeta<M>>()
+                .init_resource::<SpecializedRenderPipelines<UiMaterialPipeline<M>>>()
+                .add_systems(
+                    ExtractSchedule,
+                    (
+                        extract_ui_materials::<M>,
+                        extract_ui_material_nodes::<M>.in_set(RenderUiSystem::ExtractNode),
+                    ),
+                )
+                .add_systems(
+                    Render,
+                    (
+                        prepare_ui_materials::<M>.in_set(RenderSet::PrepareAssets),
+                        queue_ui_material_nodes::<M>.in_set(RenderSet::Queue),
+                        prepare_uimaterial_nodes::<M>.in_set(RenderSet::PrepareBindGroups),
+                    ),
+                );
+        }
+    }
+
+    fn finish(&self, app: &mut App) {
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app.init_resource::<UiMaterialPipeline<M>>();
+        }
+    }
+}
+
+#[derive(Resource)]
+pub struct UiMaterialMeta<M: UiMaterial> {
```

Comment:
Does this need to be per material or would one for all materials + instance offsets be enough?

Reply:
I don’t think so. Should I initialize this one in the normal zu initialization or is there another way to do it once?

Reply:
Apologies for the obviously bad question, I did read up on it and I updated everything. Now uses only one UI Meta (like the normal UI does)